### PR TITLE
Implement selected CIS

### PR DESCRIPTION
**Pull request by lelemmen**
**Short description**

In this PR, we’ll implement CIS through a named constructor for `SelectedSpinResolvedONVBasis`.

**Related issues**

**To do**

- [x] (if applicable, when creating new source files): don't forget to update the Doxygen file, the collective `gqcp.hpp` include header and the CMake files
- [x] Provide a test case for the full H2O spectrum
- [x] Clean up the source code.
- [x] Add a boolean argument indicating if the triplet excitations should be included

### CODE REVIEW

**Comment by lelemmen** [comment]
@AaronDC60 Would you be up for cleaning up this PR and finalizing it? I know you’re currently busy with your own CIS implementation, so I think you’ll learn a lot from this PR.

File: gqcp/src/ONVBasis/SpinResolvedSelectedONVBasis.cpp [hunk 1]
```diff
@@ -106,6 +106,117 @@ SpinResolvedSelectedONVBasis::SpinResolvedSelectedONVBasis(const SpinResolvedONV
...
+            std::cout << "a: " << a_beta << std::endl;
+            auto alpha_part = alpha_reference;
+            auto beta_part = beta_reference;
+
+            onvs.emplace_back(alpha_part, beta_part);
+        }
+    }
+
+
+    // Generate the alpha-beta excitations.
```
Comment (dariatols) [inline gqcp/src/ONVBasis/SpinResolvedSelectedONVBasis.cpp:191]:
```suggestion
    // Generate the beta-alpha excitations.
```

**Comment by dariatols** [comment]
@lelemmen The source code has been cleaned and the test case providing the full CIS spectrum for H2O has been calculated using @AaronDC60's `Psi4` calculations. We have adjusted the code to count excitations instead of the number of differences when comparing ONVs. 

Additionally, I have added a Python binding such that @ruvdrsti can use `GQCPy` for his CIS implementations with a CUHF reference as well.

File: gqcp/include/ONVBasis/SpinResolvedSelectedONVBasis.hpp [hunk 2]
```diff
@@ -378,7 +395,7 @@ class SpinResolvedSelectedONVBasis {
                 SpinUnresolvedONV beta_J = onv_J.onv(Spin::beta);
 
                 // 1 excitation in the alpha part, 0 in the beta part.
-                if ((alpha_I.countNumberOfDifferences(alpha_J) == 2) && (beta_I.countNumberOfDifferences(beta_J) == 0)) {
+                if ((alpha_I.countNumberOfExcitations(alpha_J) == 1) && (beta_I.countNumberOfExcitations(beta_J) == 0)) {
```
Comment (lelemmen) [inline gqcp/include/ONVBasis/SpinResolvedSelectedONVBasis.hpp:398]: Thanks for cleaning this up. Not only did you fix a 'wrong use of an API', but it's clearer to read as well.

File: gqcp/tests/CMakeLists.txt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 set(test_target_sources)
 
-add_subdirectory(Basis)
+#add_subdirectory(Basis)
```
Comment (lelemmen) [inline gqcp/tests/CMakeLists.txt:3]:
```suggestion
add_subdirectory(Basis)
```

Could you re-enable these tests?

File: gqcp/tests/QCMethod/CI/selected_CI_test.cpp [hunk 4]
```diff
@@ -90,3 +96,67 @@ BOOST_AUTO_TEST_CASE(unrestricted_selected_FCI) {
...
+        -73.833309015644829,
+        -73.741983861357141,
+        -73.641294789858790,
+        -73.616317921591587,
+        -54.983553526518087,
+        -54.931100523669279,
+        -54.930737857663736,
+        -54.891548004632547};
+
+    // Create the molecular Hamiltonian in the Löwdin basis.
```
Comment (lelemmen) [inline gqcp/tests/QCMethod/CI/selected_CI_test.cpp:130]:
```suggestion
    // Create the molecular Hamiltonian in the RHF orbital basis.
```

File: gqcp/tests/QCMethod/CI/selected_CI_test.cpp [hunk 5]
```diff
@@ -90,3 +96,67 @@ BOOST_AUTO_TEST_CASE(unrestricted_selected_FCI) {
...
+    const auto molecule = GQCP::Molecule::ReadXYZ("data/h2o_cis.xyz");
+
+    GQCP::RSpinOrbitalBasis<double, GQCP::GTOShell> spin_orbital_basis {molecule, "STO-3G"};
+    // spin_orbital_basis.lowdinOrthonormalize();
+    auto hamiltonian = spin_orbital_basis.quantize(GQCP::FQMolecularHamiltonian(molecule));
+    // const auto K = hamiltonian.numberOfOrbitals();
```
Comment (lelemmen) [inline gqcp/tests/QCMethod/CI/selected_CI_test.cpp:136]:
```suggestion
    auto hamiltonian = spin_orbital_basis.quantize(GQCP::FQMolecularHamiltonian(molecule));
```

File: gqcp/tests/QCMethod/CI/selected_CI_test.cpp [hunk 6]
```diff
@@ -90,3 +96,67 @@ BOOST_AUTO_TEST_CASE(unrestricted_selected_FCI) {
...
+
+    // Solve the RHF SCF equations to find an initial orthonormal basis.
```
Comment (lelemmen) [inline gqcp/tests/QCMethod/CI/selected_CI_test.cpp:138]:
```suggestion
    // Solve the RHF SCF equations.
```

File: gqcp/tests/QCMethod/CI/selected_CI_test.cpp [hunk 7]
```diff
@@ -90,3 +96,67 @@ BOOST_AUTO_TEST_CASE(unrestricted_selected_FCI) {
...
+    auto rhf_environment = GQCP::RHFSCFEnvironment<double>::WithCoreGuess(molecule.numberOfElectrons(), hamiltonian, spin_orbital_basis.overlap().parameters());
+    auto plain_rhf_scf_solver = GQCP::RHFSCFSolver<double>::Plain();
+    const GQCP::DiagonalRHFFockMatrixObjective<double> objective {hamiltonian};
+    const auto rhf_qc_structure = GQCP::QCMethod::RHF<double>().optimize(objective, plain_rhf_scf_solver, rhf_environment);
+    const auto rhf_parameters = rhf_qc_structure.groundStateParameters();
+
+    hamiltonian.transform(rhf_parameters.expansion());
+
+
+    // Set up the full spin-resolved selected ONV basis for CIS where triplet excitations are included.
```
Comment (lelemmen) [inline gqcp/tests/QCMethod/CI/selected_CI_test.cpp:148]:
```suggestion
    // Set up the spin-resolved selected ONV basis for CIS where triplet excitations are included.
```

File: gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp [hunk 8]
```diff
@@ -42,8 +42,8 @@ void bindSpinResolvedSelectedONVBasis(py::module& module) {
     py_SpinResolvedSelectedONVBasis
 
         /*
-         *  MARK: Constructors
-         */
+          *  MARK: Constructors
+          */
```
Comment (lelemmen) [inline gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp:46]:
```suggestion
         *  MARK: Constructors
         */
```

File: gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp [hunk 9]
```diff
@@ -61,8 +61,18 @@ void bindSpinResolvedSelectedONVBasis(py::module& module) {
 
 
         /*
-         *  MARK: General information
-         */
+          *  MARK: Named constructors
+          */
```
Comment (lelemmen) [inline gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp:65]:
```suggestion
         *  MARK: Named constructors
         */
```

File: gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp [hunk 10]
```diff
@@ -61,8 +61,18 @@ void bindSpinResolvedSelectedONVBasis(py::module& module) {
...
+
+        .def_static(
+            "CIS",
+            &SpinResolvedSelectedONVBasis::CIS,
+            "Return a CI singles-equivalent `SpinResolvedSelectedONVBasis`")
+
+
+        /*
+          *  MARK: General information
+          */
```
Comment (lelemmen) [inline gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp:75]:
```suggestion
         *  MARK: General information
         */
```

File: gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp [hunk 11]
```diff
@@ -61,8 +61,18 @@ void bindSpinResolvedSelectedONVBasis(py::module& module) {
...
         /*
-         *  MARK: General information
-         */
+          *  MARK: Named constructors
+          */
+
+        .def_static(
+            "CIS",
+            &SpinResolvedSelectedONVBasis::CIS,
+            "Return a CI singles-equivalent `SpinResolvedSelectedONVBasis`")
```
Comment (lelemmen) [inline gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp:70]:
```suggestion
            "Return a CI singles-equivalent `SpinResolvedSelectedONVBasis`.")
```

File: gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp [hunk 12]
```diff
@@ -82,8 +92,8 @@ void bindSpinResolvedSelectedONVBasis(py::module& module) {
 
 
         /*
-         *  MARK: Modifying
-         */
+          *  MARK: Modifying
+          */
```
Comment (lelemmen) [inline gqcpy/src/ONVBasis/SpinResolvedSelectedONVBasis_bindings.cpp:96]:
```suggestion
         *  MARK: Modifying
         */
```

**Review by lelemmen — COMMENTED** [review]
In this review, I've looked at the structure of the code and not necessarily checked its correctness. I trust that you've added the necessary test cases.

**Review by dariatols — APPROVED** [review]

**Pull request merged**